neo.UnitTests/Ledger/UT_MemoryPool.cs

```diff
@@ -253,5 +287,205 @@ public void TestInvalidateAll()
+            _unit.InvalidateVerifiedTransactions();
+            AddTransactions(513);
+            var result = _unit.ReVerifyTopUnverifiedTransactionsIfNeeded(10, Blockchain.Singleton.GetSnapshot());
+            _unit.VerifiedCount.Should().Be(514);
+            _unit.UnVerifiedCount.Should().Be(1);
+            result.Should().BeTrue();
+        }
+
+        [TestMethod]
+        public void TestTryAdd()
+        {
+            var tx1 = CreateTransaction();
+            _unit.TryAdd(tx1.Hash, tx1);
+            _unit.TryAdd(tx1.Hash, tx1).Should().BeFalse();
+            _unit2.TryAdd(tx1.Hash, tx1).Should().BeFalse();
+        }
+
+        [TestMethod]
+        public void TestTryGetValue()
+        {
+            var tx1 = CreateTransaction();
+            _unit.TryAdd(tx1.Hash, tx1);
+            _unit.TryGetValue(tx1.Hash, out Transaction tx).Should().BeTrue();
+            tx.ShouldBeEquivalentTo(tx1);
+
+            _unit.InvalidateVerifiedTransactions();
+            _unit.TryGetValue(tx1.Hash, out tx).Should().BeTrue();
+            tx.ShouldBeEquivalentTo(tx1);
+
+            var tx2 = CreateTransaction();
+            _unit.TryGetValue(tx2.Hash, out tx).Should().BeFalse();
+        }
+
+        [TestMethod]
+        public void TestUpdatePoolForBlockPersisted()
+        {
+            var mockSnapshot = new Mock<Snapshot>();
+            byte[] transactionsPerBlock = { 0x18, 0x00, 0x00, 0x00 };
+            byte[] feePerByte = { 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00 };
+            StorageItem item1 = new StorageItem
+            {
+                Value = transactionsPerBlock
+            };
+            StorageItem item2 = new StorageItem
+            {
+                Value = feePerByte
+            };
+            var myDataCache = new MyDataCache<StorageKey, StorageItem>();
+            var key1 = CreateStorageKey(Prefix_MaxTransactionsPerBlock);
+            var key2 = CreateStorageKey(Prefix_FeePerByte);
+            var ServiceHash = "Neo.Native.Policy".ToInteropMethodHash();
+            byte[] script = null;
+            using (ScriptBuilder sb = new ScriptBuilder())
+            {
+                sb.EmitSysCall(ServiceHash);
+                script = sb.ToArray();
+            }
+            var Hash = script.ToScriptHash();
+            key1.ScriptHash = Hash;
+            key2.ScriptHash = Hash;
+            myDataCache.Add(key1, item1);
+            myDataCache.Add(key2, item2);
+            mockSnapshot.SetupGet(p => p.Storages).Returns(myDataCache);
+
+            var tx1 = CreateTransaction();
+            var tx2 = CreateTransaction();
+            var tx3 = CreateTransaction();
+            Transaction[] transactions = { tx1, tx2 };
+            _unit.TryAdd(tx1.Hash, tx1);
+            _unit.TryAdd(tx3.Hash, tx3);
+            var block = new Block { Transactions = transactions };
+
+            _unit.UnVerifiedCount.Should().Be(0);
+            _unit.VerifiedCount.Should().Be(2);
```

Comment:
Block had transactions `tx1` and `tx2`, right?
`tx3` was just in the mempool?

Reply:
Yes, tx1 is in both memory pool and block, tx2 is just in block and tx3 is just in memory pool.
tx3 is used to test the case that the memory pool should clean up all the transactions in it when `UpdatePoolForBlockPersisted` is called.

Reply:
I got it, @eryeer. 
However, it looks strange...

Why is `tx3` cleaned if it was in the mempool? Just `tx1` and `tx2` were persisted.

The method `UpdatePoolForBlockPersisted` should had left `tx3`, no?

Reply:
tx3 is to describe the case that if `policyChanged == true` in Mempool, all the rest of the transactions in the Mempool should be removed. The method `UpdatePoolForBlockPersisted` in line 428 takes a different snapshot and the policy is changed, so tx3 should be removed.

Reply:
Thanks for the explanation.

I will soon take another look, because I remember that we were trying to enforce the removal only if fees are not anymore valid.

Reply:
@eryeer, It is still strange do me, even if `PolicyChanged` it should not clean tx3 because its `fee` is enough for the current modified policy, right?

Maybe there is an error on the method `ReverifyTransactions`, which should reverify `tx3`back to the mempool.

Or, perhaps, we need to cache the tx's and then send to system.Blockchain.Tell.
```cs
                    foreach (PoolItem item in _unverifiedSortedTransactions.Reverse())
                        if (item.Tx.FeePerByte >= _feePerByte)
                            _system.Blockchain.Tell(item.Tx, ActorRefs.NoSender);
                    _unverifiedTransactions.Clear();
                    _unverifiedSortedTransactions.Clear();
```

Something like:

```cs
                    foreach (PoolItem item in _unverifiedSortedTransactions.Reverse())
                        if (item.Tx.FeePerByte >= _feePerByte)
                            cache item.TX into a vector itemsCached
                    _unverifiedTransactions.Clear();
                    _unverifiedSortedTransactions.Clear();

                    foreach (Transaction item in itemsCached)
                            _system.Blockchain.Tell(items[].Tx, ActorRefs.NoSender);
```

Reply:
@erikzhang, can you please double check this behavior?